File: GRPC/GrpcTodoList/Services/TodoListService.cs
using Grpc.Core;
using GrpcTodoList;

namespace GrpcTodoList
{
    public class TodoListService : TodoList.TodoListBase
    {
        static private int TodoItem_Next_ID = 0;
        static private List<TodoItem> Datas = new List<TodoItem>();

        private readonly ILogger<TodoListService> _logger;

        public TodoListService(ILogger<TodoListService> logger)
        {
            _logger = logger;

            if (Datas == null)
                InitialiszeDefaultData();
        }

        private void InitialiszeDefaultData()
        {
            Datas = new List<TodoItem>();
            int iNbItemToCreate = 5;
            for (int i = 1; i < iNbItemToCreate; i++)
            {
                TodoItem_Next_ID++;
                TodoItem data = new TodoItem
                {
                    Id = TodoItem_Next_ID,
                    Titre = $"Titre test {TodoItem_Next_ID}",
                    Description = $"Description test {TodoItem_Next_ID}"
                };
                Datas.Add(data);
            }
        }

        /// <summary>
        /// Creation d'un item de todo
        /// </summary>
        /// <param name="request"></param>
        /// <param name="context"></param>
        /// <returns></returns>
        public override Task<TodoItem> CreateTodoItem(CreateTodoItemRequest request, ServerCallContext context)
        {
            TodoItem_Next_ID++;
            TodoItem item = new TodoItem 
                                {
                                    Id = TodoItem_Next_ID,
                                    Titre = request.Titre,
                                    Description = request.Description,
                                };
            Datas.Add(item);
            return Task.FromResult(item);
        }

        /// <summary>
        /// Lecture d'un todo item
        /// </summary>
        /// <param name="request"></param>
        /// <param name="context"></param>
        /// <returns></returns>
        public override Task<TodoItem> GetTodoItem(GetTodoItemRequest request, ServerCallContext context)
        {
            TodoItem? result;
            if (Datas.Count > 0)
            {
                TodoItem? item = Datas.FirstOrDefault(i => i.Id == request.Id);
                if (item == null)
                    result = new TodoItem();
                else
                    result = new TodoItem
                    {
                                    Id = item.Id,
                                    Titre = item.Titre,
                                    Description = item.Description
                                };
            }
            else
                result = new TodoItem(); // Todo item vide

            return Task.FromResult(result);
        }

        /// <summary>
        /// Demande la listes des todo items
        /// </summary>
        /// <param name="request"></param>
        /// <param name="context"></param>
        /// <returns></returns>
        public override Task<GetTodoListReply> GetTodoList(GetTodoListRequest request, ServerCallContext context)
        {
            // recherche a faire

            // TODO : Renvoyer une liste des todo list
            var reply = new GetTodoListReply();
            reply.Items.AddRange(Datas);
            return Task.FromResult(reply);
        }
    }
}

File: GRPC-java/GrpcTodoListServer/src/main/java/todo/TodoListService.java
package todo;

import java.util.ArrayList;
import java.util.List;

import todo.Todo.TodoItem;
import todo.TodoListGrpc.TodoListImplBase;

public class TodoListService extends TodoListImplBase {
	
	/**
	 * La liste est maintenue en mémoire pour l'exemple
	 */
	private List<TodoRecord> todoList = new ArrayList<>();
	
    /**
     * Creation d'un item
     */
    @Override
    public void createTodoItem(todo.Todo.CreateTodoItemRequest request,
        io.grpc.stub.StreamObserver<todo.Todo.TodoItem> responseObserver) {
      System.out.println("createTodoItem");
      var titre = request.getTitre();
      var description = request.getDescription();
      todoList.add(new TodoRecord(titre, description));
      responseObserver.onNext(TodoItem
              .newBuilder()
              .setTitre(titre)
              .setDescription(description)
              .build());
      responseObserver.onCompleted();
    }

    /**
     * Lecture d'un item
     */
    @Override
    public void getTodoItem(todo.Todo.GetTodoItemRequest request,
        io.grpc.stub.StreamObserver<todo.Todo.TodoItem> responseObserver) {      
      System.out.println("getTodoItem");
      var id = request.getId();
      var todo = todoList.get(id);
      responseObserver.onNext(TodoItem
              .newBuilder()
              .setTitre(todo.titre())
              .setDescription(todo.description())
              .build());
      responseObserver.onCompleted();
    }

    /**
     * Demande la liste des items
     */
    @Override
    public void getTodoList(com.google.protobuf.Empty request,
        io.grpc.stub.StreamObserver<todo.Todo.GetTodoListReply> responseObserver) {
        System.out.println("getTodoList");
        var todoBuilder = todo.Todo.GetTodoListReply
                .newBuilder();
        for(var todo : todoList) {
            todoBuilder.setItems(0, TodoItem
            		.newBuilder()
            		.setTitre(todo.titre())
            		.setDescription(todo.description())
            		.build());
        }
        responseObserver.onNext(todoBuilder
                .build());
        responseObserver.onCompleted();
    }

}
